### ISSUE TITLE
Problem on MSeg, took 10 minutes for 1 image and only produced the gray scaled image?

### ISSUE THREAD

**Issue by luda1013** [isl-org/PhotorealismEnhancement#56]
Hallo Pros,

i am currently working with enhancing image enhancement paper and algorithm and trying to implement that. In the process, we need to use MSeg-segmentation for real and rendered images/ datasets. i have like 50-60k images.

So the dependencies MSeg-api and MSeg_semantic were already installed. I tried the google collab first and then copying the commands, so i could run the script in my linux also. the command is like this:
python -u mseg_semantic/tool/universal_demo.py \
  --config="default_config_360.yaml" \
  model_name mseg-3m \
  model_path mseg-3m.pth \
  input_file /home/luda1013/PfD/image/try_images

the weight i used, i downloaded it from the google collab, so the mseg-3m-1080.pth
![MSeg-log](https://user-images.githubusercontent.com/70270770/233037270-5c666568-ede6-4f85-8bf0-85ee97ec3a10.PNG)

but for me, it took like 10 minutes for 1 image and also what i get in temp_files is just the gray scale image of it.
Could someone help me how i could solve this problem, thank you :)

**Comment by luda1013** [comment]
My setup:
- Ubuntu 20.04.6 LTS
- core i9-10980XE @ 3GHz
- graphic (nvidia-smi):  NVIDIA RTX A5000 (48 GB)
- Cuda 11.7
- Pytorch version : 2.0.0
-  Cuda at: /usr/local/cuda*

**Comment by czero69** [comment]
Sounds like you are running on cpu and not on gpu. I am runninng on 24gb gpu and for "default_config_1080_ms.yaml" config it took good few seconds for one fullHD input image. Check in your config what gpu is indicated as test_gpu: [0]. Make sure also pytorch is picking nvidia as a device and not a cpu. It is the often case where you have something badly installed in your environment. From log lines you should indicate or you can also write super simple pytorch script to print device name (google for it). If it is the case, I would create new conda environment and install pytorch for gpu/cuda (pip install might be bit different for gpu support).

About grayscale output it is probably what you want. One channel (8 bit gray) is super enough to store 256 classes and mseg is producing less than that. This way of storing info is efficient for hdd, 4k image is barely 112kb or so.

**Comment by luda1013** [comment]
ok, thank you for the reply.. ok,  i will make new virtual environment just for the MSeg. about the cpu and gpu, when running the script, i also check  the nvidia-smi and it is on 100% load
![Nvidia when MSeg active](https://github.com/isl-org/PhotorealismEnhancement/assets/70270770/2f61c7c7-e271-4293-b5a3-3ffe21bba50b)

is okay to use another segmentation network for this enhancing photorealism enhancement?

**Comment by luda1013** [comment]
okay, so the MSeg scipt i used is the universal demo inference  or the universal demo for batch; so the command is like:
python -u ~/mseg-semantic/mseg_semantic/tool/universal_demo_batched.py \
  --config=mseg_semantic/config/test/480/default_config_batched_ss.yaml\
  model_name mseg-3m \
  model_path ~/mseg-semantic/mseg-3m-480p.pth \
  input_file /home/luda1013/PfD/image/01_images/

the input file 01_images is a folder contains of 2500 images of Playing for Data - dataset
here are some screenshots on running MSeg:
![image](https://github.com/isl-org/PhotorealismEnhancement/assets/70270770/dd725137-0464-4b2d-82d6-3579d9a23cb3)
 o ya, also i dont nderstand that much about batch universal demo from Mseg, because i thought with batch, the MSeg run on batches of images, but in that command line above is like 1 batch consists only 1 image, so there are 2500 batches.

And before, i tried to use the universal_demo.py and at the first image i break the running and i think these are why it takes long time just to infer 1 image. But i dont yet investigate/debug line-per-line further.
![MSeg-log_took long time here](https://github.com/isl-org/PhotorealismEnhancement/assets/70270770/0e541282-808e-412a-ab6d-a1496b5eafac)
![MSeg-log_took long time here_2](https://github.com/isl-org/PhotorealismEnhancement/assets/70270770/0c36e1ba-b09e-4b1b-848f-f87813b9bc01)


because took too long for me for the mseg, now i have MSeg-BW-images for 1,5-2 folders of PfD: images_01 and images_02.
Now i am using it to implements the EPE, 1 folder as fake , 2nd folder as real dataset.
Have you tried also to implement and train the EPE? because till now i still can't bring it for training.
i have error when run the training:
![image](https://github.com/isl-org/PhotorealismEnhancement/assets/70270770/d3690650-8010-49eb-a0e3-72a874cf72d6)

![image](https://github.com/isl-org/PhotorealismEnhancement/assets/70270770/26049c38-0fb8-4ba5-b704-0da5ccb82be3)

now i am still at debug line per line where this num_samples = 0 came..

Thank you btw for your  help

**Comment by czero69** [comment]
> o ya, also i dont nderstand that much about batch universal demo from Mseg, because i thought with batch, the MSeg run on batches of images, but in that command line above is like 1 batch consists only 1 image, so there are 2500 batches.

ah, I am going only with universal_demo.py

> i have error when run the training:
you have at least two errors. One says some file is missing. Another one, num_samples=0 this happens when dataloader see no data at all, usually wrong patches / wrong input file with data.

For preparing EPE data, you must take extra care. Go throu all preparation steps careffuly. I recommend printing results of every step below (values, means) to check does tensors looks ok (not nans/ not inf etc.) 

these are my scripts, where images are 4k (hence 2160 3840 and -c 60)
```
python epe/matching/feature_based/collect_crops.py FirstRealVids /path/real.txt -n 8 -c 60
python epe/matching/feature_based/collect_crops.py Coffing /path/fake.txt -n 8 -c 60
python epe/matching/feature_based/find_knn.py crop_Coffing.npz crop_FirstRealVids.npz knn_Coffing-FirstRealVids.npz -k 15
python epe/matching/filter.py knn_Coffing-FirstRealVids.npz crop_Coffing.csv crop_FirstRealVids.csv 1.0 matched_crops_Coffing-FirstRealVids.csv
python epe/matching/compute_weights.py matched_crops_Coffing-FirstRealVids.csv 2160 3840 crop_weights_Coffing-FirstRealVids.npz
```

take note of correct order for /path/real.txt (images, msegs)
and for /path/fake.txt (images, msegs, gbuffers_npz, gt_stencils)

To verify a bit, its good to see does matched crops looks ok
```
python ./epe/matching/feature_based/sample_matches.py  /path/fake.txt crop_Coffing.csv /path/real.txt crop_FirstRealVids.csv knn_Coffing-FirstRealVids.npz
```

Also, make sure all your input color images are RGB, 3-channel (not RGBA). Robust maps (mseg) and Stencils (gt masks) are in 8 bit. Your NPZs should have a same structure as fake NPZs ('data' key in numpy dict, float16). (If it has different dim than 32 modify in code accordingly, should be in one place)

> Have you tried also to implement and train the EPE? because till now i still can't bring it for training.

After solving some trivial issues, all is training fine for me. Results are ... shortly speaking ... breathtaking. Possibly I would rewrite an entire training pipeline to latest pytorch and pipeline similar to how I work nowadays, so it would be easier for me to modify epe basilne arch furhter, support batches > 1, logging, etc.

**Comment by czero69** [comment]
in compute_weights.py take a note that argument is in H, W (and not W, H), so for e.g. fullHD will be 1080 1920; it was my NaN reason

**Comment by luda1013** [comment]
oww.. so the txt file i created is false? i thought is correct because i followed the instructions 
" Each line should contain paths to image, robust label map, gbuffer file, and ground truth label map, all separated by commas.  Each line should contain paths to image, robust label map, gbuffer file, and ground truth label map, all separated by commas."

sorry but i also dont quite understand with the stat in the terminal, so u suggest to verify the location of the image using stat?
![image](https://github.com/isl-org/PhotorealismEnhancement/assets/70270770/2b422537-8846-4cc8-98bb-19b72f996292)

maybe could help u too look better, here is the part of screen shot of my rendered.txt:
![image](https://github.com/isl-org/PhotorealismEnhancement/assets/70270770/49032c42-c236-4a9a-8576-b9171a522b58)

**Comment by vace17** [comment]
Hello @luda1013,
I noticed you are using CUDA version 11.7, were you able to train the model properly?                                                                                          
I was also using this version of CUDA but the training step takes a lot of time, I supposed because pytorch is not using the GPU (maybe because of incompatible CUDA?) 
Thank you in advance for any help

**Comment by luda1013** [comment]
Hi @vace17 , not yet, i am still countered some problem, now i countered the problem:
![error_train](https://github.com/isl-org/PhotorealismEnhancement/assets/70270770/c15e4d54-240b-4b88-a01e-dab36e5a2fb9)
seems that i need to squeeze or reshape the image first into (H,W) instead (H,W,C)

in your case, in the script they always made the used device is cuda, u can also then check when u are on training, open your terminal and check with nvidia-smi
u should then know if your gpu is loaded or not

Could you please help me then with training? till now i cannot bring it to train.
may i know your dataset for fake and real, the config and the step by step? do u just train or did u also modify the scripts? thank you very much :)

**Comment by vace17** [comment]
@luda1013 I have a different issue since I don't encounter this specific error and the training process is running but it takes a lot of time

**Comment by vace17** [comment]
@czero69 can I ask what is your specific setup?
Mine is :

- Windows 10
- Graphic card: NVIDIA GEFORCE RTX 3090 
- Cuda 11.7
- Python version 3.8.16

I checked from the terminal the current usage of the GPU using the command nvidia-smi during the run of the training process.   It seems to me that the GPU is currently used but the percentage of usage of it continues changing between low values (5-10%) and 50-60%.
How much time does it take to train the framework with your current setup?
Thank you in advance for any help

**Comment by czero69** [comment]
hey, I have tried two set-ups so far:

- win10, NVIDIA RTX 3090 it trained 75k epoch / day
- linux, NVIDIA A6000 (48GB), ~200k epoch / day

Authors mentioned somewhere here in the issue space here that for them it was around 200k epochs too

**Comment by ZGX010** [comment]
> > oww.. so the txt file i created is false?
> 
> at least order looks ok. I have this order too ["screenshot", "msegs/gray4k", "NPZs", "gray_stencils"]
> 
> the stat is just to make sure .txt/.csv have correct paths. The one you've printed is correct, indicating file indeed exist in HDD. Rendered.txt looks ok too for a first sight. But num_samples == 0 indicates epe pipeline does not see sth. Check all 4 paths in some random row in rendered.txt. Check paths in your citysample_ie2.config (or whatever name config you are using for epe training). Also check whats going on with, missing files, probably some paths are pointing to non-existing files.

Hello Kamil, nice to see you again. 
gbuffer seems to be in npz format, how to organize them? 
How should the Gbuffer attributes corresponding to size and channel be defined?
